[WebDriver] If Safari window is not focused, WebDriver-automated mouse events do not get sent to the page

the-chenergy opened the pull request:
#### 3e29d37be2b6cd98183fd643223f0070f55c913e
<pre>
[WebDriver] If Safari window is not focused, WebDriver-automated mouse events do not get sent to the page
<a href="https://rdar.apple.com/117035696">rdar://117035696</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=275326">https://bugs.webkit.org/show_bug.cgi?id=275326</a>

Reviewed by Patrick Angle.

When the Safari window is not currently focused, AKA not the key window
using AppKit&apos;s terminology, safaridriver-automated mouse events get
sent to the window but not to the webpage within it.

The patch <a href="https://github.com/WebKit/WebKit/commit/f41c94d11e5a0b8900cf768adeb76d94ef681444#diff-8eae0c0334ee2210a82eef8240463d484f408aceba25aef9f96093c9ca042863">https://github.com/WebKit/WebKit/commit/f41c94d11e5a0b8900cf768adeb76d94ef681444#diff-8eae0c0334ee2210a82eef8240463d484f408aceba25aef9f96093c9ca042863</a>
attempted to make the window key, but according to AppKit engineers,
making the window key requires the application to currently be active.
Unfortunately, neither `[NSApplication.sharedApplication activate]` nor
`[NSApplication.sharedApplication activateIgnoringOtherApps:YES]`
actually guarantee activating the app immediately. This results in us
failing to ensure that the window is key before sending the mouse
events, so the mouse events could not get delivered.

   - For `mousemove` events, there is the patch <a href="https://github.com/WebKit/WebKit/pull/1431">https://github.com/WebKit/WebKit/pull/1431</a>
     that make the delivery bypass the window level and sent directly
     into the web view. However, a more recent enhancement <a href="https://github.com/WebKit/WebKit/pull/17145">https://github.com/WebKit/WebKit/pull/17145</a>
     deliberately prevented `mousemove` events from being sent to the
     web page if the page is not active, which effectively nullified the
     former patch for `mousemove` events.

To work around that, a simple solution I found was to give automated
mouse events privileges to bypass certain filtering checks, rather than
always forcing the window to steal input focus. Automated mouse events
are assigned a special, &quot;magic&quot; event number (0), so we can recognize
them in the places below in the code.

I have tested the code
   (1) Manually with Python code using Selenium launching the newly
       compiled safaridriver and seeing that mouse events always get
       delivered successfully
   (2) With `wpt run safari webdriver/tests/classic/element-click webdriver/tests/classic/element_send_keys`
       and seeing that the result are the same before and after my
       changes with the same passes and failures.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::WebAutomationSession::wasEventSynthesizedForAutomation):
(WebKit::WebAutomationSession::platformSimulateMouseInteraction):
   - Make the magic event number accessible in other places needing to
     check if an event was synthesized.

* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::acceptsFirstMouse):
   - If the mouse event is automated, bypass the check to let the window
     always accept it.
   - I don&apos;t fully understand why this works -- this is a suggestion
     from an AppKit engineer, and I thought this function was meant to
     return whether the mouse event should be deemed as &quot;activating the
     window&quot; versus &quot;activating the window and sent to the contents&quot;,
     but apparently it&apos;s just whether the event gets sent to the
     contents because automated events still don&apos;t activate the window
     regardless the function&apos;s return value.

* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::handleMouseEvent):
   - If the mouse event is automated, don&apos;t use the enhancement of
     sending the event straight to the scrollbars and still send it
     to the web page.

* Tools/TestWebKitAPI/Tests/mac/AcceptsFirstMouse.mm:
(TestWebKitAPI::AcceptsFirstMouse::runTest):
   - Creating the event with eventNumber:0 for testing resulted in the
     event being considered as synthesized and acceptsFirstMouse always
     returned YES. Use a different eventNumber instead as a workaround.

Canonical link: <a href="https://commits.webkit.org/280047@main">https://commits.webkit.org/280047@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1433b9f10bd5c5fca313d152591cbc6517f95b75

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34888 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8032 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58549 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5995 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57691 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42510 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6194 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44747 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4116 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57594 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32793 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47891 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25874 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29577 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5220 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4139 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51510 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5487 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60140 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30719 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5611 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52178 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31804 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47961 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51658 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12315 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32885 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31551 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -364,6 +364,10 @@ static inline bool expectsLegacyImplicitRubberBandControl()
 
 bool WebPageProxy::acceptsFirstMouse(int eventNumber, const WebKit::WebMouseEvent& event)
 {
+    const int synthesizedMouseEventMagicEventNumber = 0;
```
If we are going to start using this outside of WebAutomationSessionMac we probably want to to instead move the declaration of the original static constant to a header like `WebAutomationSession.h` so we only declare it once, and give is a new name to make it clear it is used for automation, since that isn't clear here. Maybe `webAutomationSynthesizedMouseEventMagicEventNumber`

patrickangle commented on Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
```diff
@@ -101,9 +101,6 @@
 
-        // Take focus back in case the Inspector became focused while the prior command or
-        // NSEvent was delivered to the window.
-        [window makeKeyAndOrderFront:nil];
```
Is removing this necessary. It still seems like the original comment would apply if the application is active but the window we are targeting isn't front and center yet?

the-chenergy replied:
Correct, removing this is not necessary. Also, given that my fix doesn't address the bug with modifier-included keyboard events, we could put this back so we at least make an attempt to focus the window so that those events have a chance to work.

patrickangle reviewed (approved):
We should keep looking in to how we can get the system to provide the same behavior it used to, where makeKeyAndOrderFront would work for a backgrounded app, but this seems like a reasonable workaround in the meantime. We should open a bug to that affect and put a FIXME with the bug number above the workarounds in WebPageProxyMac and WebFrame.

the-chenergy committed:
[WebDriver] If Safari window is not focused, WebDriver-automated mouse events do not get sent to the page
rdar://117035696
https://bugs.webkit.org/show_bug.cgi?id=275326

Reviewed by Patrick Angle.

When the Safari window is not currently focused, AKA not the key window
using AppKit's terminology, safaridriver-automated mouse events get
sent to the window but not to the webpage within it.

The patch https://github.com/WebKit/WebKit/commit/f41c94d11e5a0b8900cf768adeb76d94ef681444#diff-8eae0c0334ee2210a82eef8240463d484f408aceba25aef9f96093c9ca042863
attempted to make the window key, but according to AppKit engineers,
making the window key requires the application to currently be active.
Unfortunately, neither `[NSApplication.sharedApplication activate]` nor
`[NSApplication.sharedApplication activateIgnoringOtherApps:YES]`
actually guarantee activating the app immediately. This results in us
failing to ensure that the window is key before sending the mouse
events, so the mouse events could not get delivered.

   - For `mousemove` events, there is the patch https://github.com/WebKit/WebKit/pull/1431
     that make the delivery bypass the window level and sent directly
     into the web view. However, a more recent enhancement https://github.com/WebKit/WebKit/pull/17145
     deliberately prevented `mousemove` events from being sent to the
     web page if the page is not active, which effectively nullified the
     former patch for `mousemove` events.

To work around that, a simple solution I found was to give automated
mouse events privileges to bypass certain filtering checks, rather than
always forcing the window to steal input focus. Automated mouse events
are assigned a special, "magic" event number (0), so we can recognize
them in the places below in the code.

I have tested the code
   (1) Manually with Python code using Selenium launching the newly
       compiled safaridriver and seeing that mouse events always get
       delivered successfully
   (2) With `wpt run safari webdriver/tests/classic/element-click webdriver/tests/classic/element_send_keys`
       and seeing that the result are the same before and after my
       changes with the same passes and failures.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::WebAutomationSession::wasEventSynthesizedForAutomation):
(WebKit::WebAutomationSession::platformSimulateMouseInteraction):
   - Make the magic event number accessible in other places needing to
     check if an event was synthesized.

* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::acceptsFirstMouse):
   - If the mouse event is automated, bypass the check to let the window
     always accept it.
   - I don't fully understand why this works -- this is a suggestion
     from an AppKit engineer, and I thought this function was meant to
     return whether the mouse event should be deemed as "activating the
     window" versus "activating the window and sent to the contents",
     but apparently it's just whether the event gets sent to the
     contents because automated events still don't activate the window
     regardless the function's return value.

* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::handleMouseEvent):
   - If the mouse event is automated, don't use the enhancement of
     sending the event straight to the scrollbars and still send it
     to the web page.

* Tools/TestWebKitAPI/Tests/mac/AcceptsFirstMouse.mm:
(TestWebKitAPI::AcceptsFirstMouse::runTest):
   - Creating the event with eventNumber:0 for testing resulted in the
     event being considered as synthesized and acceptsFirstMouse always
     returned YES. Use a different eventNumber instead as a workaround.

Canonical link: https://commits.webkit.org/280047@main
